stan/math/rev/functor/operands_and_partials.hpp

```diff
@@ -36,14 +36,44 @@ class ops_partials_edge<double, var> {
  private:
   template <typename, typename, typename, typename, typename, typename>
   friend class stan::math::operands_and_partials;
-  const var& operand_;
-
-  void dump_partials(double* partials) { *partials = this->partial_; }
-  void dump_operands(vari** varis) { *varis = this->operand_.vi_; }
-  int size() const { return 1; }
-  std::tuple<> container_operands() { return std::tuple<>(); }
-  std::tuple<> container_partials() { return std::tuple<>(); }
+  var operand_;
+  static constexpr int size() noexcept { return 1; }
+  inline auto operand() noexcept { return this->operand_; }
+  inline auto partial() noexcept { return this->partial_; }
 };
+
+template <typename T1, typename T2,
+          require_all_kernel_expressions_and_none_scalar_t<T1, T2>* = nullptr>
+inline void update_adjoints(var_value<T1>& x, T2&& y, const var& z) {
+  x.adj() += z.adj() * y;
+}
+
+template <typename Scalar1, typename Scalar2, require_var_t<Scalar1>* = nullptr,
+          require_not_var_matrix_t<Scalar1>* = nullptr,
+          require_arithmetic_t<Scalar2>* = nullptr>
+inline void update_adjoints(Scalar1& x, Scalar2&& y, const var& z) {
+  x.adj() += z.adj() * y;
+}
```

Comment:
I'm passing these scalars by value now :-)

Though actually if we are we are passing around a `var_value`, an `arena_t<>` type, and a var, could these both (all) be just passing by value?

Reply:
Yes!

Reply:
I'm trying to think on the logic of why it's always better to pass these by value. Is it just that when we have `foo(T& x)` where x holds only a pointer (and for arena matrices a pointer and some integers) if they are going to be loaded and used into this function anyway it's better to have `foo(T x)` since the original becomes something like a pointer to a pointer + some ints where it's cheaper just to copy the pointer and ints.

I wonder if that difference is even benchmarkable though?